### PR TITLE
e2e test - ConfigMap updates

### PR DESCRIPTION
**Pull request by jsanda**
This is a WIP and **not** yet ready for merge. I wanted to go ahead and open the PR to share what I am doing and see if there is any feedback.

Currently the test only uses a one DC and one rack. I plan to update the test to use multiple DCs and multiple racks. I would also like to add more fine-grained checking of status updates to hopefully reduce timeouts.

### CODE REVIEW

**Comment by jsanda** [comment]
This is ready for review. This PR adds a test for config map updates within a single DC. I thought it might be good to do a multi-DC test in separate function that I could submit as follow up PR. 

I have seen a couple failures with this error:

```
wait_util.go:211: Waiting for full Operator configmap-upgrade to finish Action of UpdateConfigMap=Ongoing
...
cassandra-e2e   3m37s       Warning   FailedPreStopHook        Pod                   Exec lifecycle hook ([/bin/bash -c /etc/cassandra/pre_stop.sh]) for Container "cassandra"
...
Jul 07, 2019 5:26:43 AM ClientCommunicatorAdmin restart
WARNING: Failed to restart: java.io.IOException: Failed to get a RMI stub: javax.naming.CommunicationException [Root exception is java.rmi.UnmarshalException: error unmarshalling return; nested exception is:
      java.io.WriteAbortedException: writing aborted; java.io.NotSerializableException: javax.management.remote.rmi.RMIJRMPServerImpl]
Jul 07, 2019 5:26:44 AM RMIConnector RMIClientCommunicatorAdmin-doStop
WARNING: Failed to call the method close():java.rmi.ConnectIOException: error during JRMP connection establishment; nested exception is:
      java.net.SocketException: Connection reset
error: no such object in table
-- StackTrace --
java.rmi.NoSuchObjectException: no such object in table
                at sun.rmi.transport.StreamRemoteCall.exceptionReceivedFromServer(StreamRemoteCall.java:283)
                at sun.rmi.transport.StreamRemoteCall.executeCall(StreamRemoteCall.java:260)
                at sun.rmi.server.UnicastRef.invoke(UnicastRef.java:161)
                at com.sun.jmx.remote.internal.PRef.invoke(Unknown Source)
                at javax.management.remote.rmi.RMIConnectionImpl_Stub.invoke(Unknown Source)
                at javax.management.remote.rmi.RMIConnector$RemoteMBeanServerConnection.invoke(RMIConnector.java:1020)
                at javax.management.MBeanServerInvocationHandler.invoke(MBeanServerInvocationHandler.java:298)
                at com.sun.proxy.$Proxy21.getDatacenter(Unknown Source)
                at org.apache.cassandra.tools.NodeTool.getOwnershipByDc(NodeTool.java:380)
                at org.apache.cassandra.tools.nodetool.Status.execute(Status.java:88)
                at org.apache.cassandra.tools.NodeTool$NodeToolCmd.run(NodeTool.java:255)
                at org.apache.cassandra.tools.NodeTool.main(NodeTool.java:169)
```
Has anyone seen this. I have not had a chance to investigate yet.

**Comment by allamand** [comment]
@jsanda, seams my script didn't catch errors, and circleCI don't report tests as KO :(

but there are some problems on script execution : https://circleci.com/gh/Orange-OpenSource/cassandra-k8s-operator/581?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link but this is not only your's seams that there is a bigger problem.

last test ok was a week ago https://circleci.com/gh/Orange-OpenSource/cassandra-k8s-operator/473

**Comment by jsanda** [comment]
Are the CircleCI errors in reference to #42? I see `CreateOneClusterService` as having failed. I added that test in #42. It looks like the operator failed to deploy. 

Interestingly though, it looks like everything passed in #42.

**Comment by allamand** [comment]
In https://circleci.com/gh/Orange-OpenSource/cassandra-k8s-operator/579?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link open the **operator acceptance test** and we see that it fails even if circleci marks the test as successful :(

**Comment by allamand** [comment]
@jsanda for the errors in the /etc/cassandra/pre_stop.sh script, may not be completely anormal, since we can have race condition between the time to execute commands in the script and the time let by kubernetes before killing the pod. This script may also be tunned in the futur

**Comment by jsanda** [comment]
@allamand hey, I took a look at `pre_stop.sh`. I don't want to deviate off topic from the PR, but I recommend some changes to the script. `nodetool drain` does all of the following (not necessarily in this order):

* stops thrift server
* stops native transport server
* stops gossip
* stops `MessagingService` (i.e., will not receive mutations from other nodes)
* stops compactions
* flush system tables
* shutdown hints 
* recycle commit log segments to avoid replay on subsequent startup

`nodetool drain` and `kill <cassandra pid>` should be sufficient for the shutdown and hopefully that will take care of the RMI exceptions.

File: test/e2e/cassandracluster_test.go [hunk 1]
```diff
@@ -2,13 +2,19 @@ package e2e
...
 	"testing"
+	"time"
 
-	framework "github.com/operator-framework/operator-sdk/pkg/test"
-	"github.com/stretchr/testify/assert"
 	"github.com/Orange-OpenSource/cassandra-k8s-operator/pkg/apis"
 	api "github.com/Orange-OpenSource/cassandra-k8s-operator/pkg/apis/db/v1alpha1"
 	mye2eutil "github.com/Orange-OpenSource/cassandra-k8s-operator/test/e2eutil"
+	framework "github.com/operator-framework/operator-sdk/pkg/test"
+	log "github.com/sirupsen/logrus"
```
Comment (cscetbon) [inline test/e2e/cassandracluster_test.go]: We use `logrus` everywhere else, we should keel doing it here to be consistent.

**Comment by allamand** [comment]
> `nodetool drain` and `kill <cassandra pid>` should be sufficient for the shutdown and hopefully that will take care of the RMI exceptions.

I think we could try that in another PR

**Comment by allamand** [comment]
@jsanda, @cscetbon actually, all tests done from a PR from fork are KO because docker image is not pushed to dockerHub, so tests can't find it. I'll work on separate PR to fix that

**Comment by allamand** [comment]
This needs to be rebased on master for tests triggering

**Comment by jsanda** [comment]
@allamand has something failed or is the build stuck?

File: test/e2e/cassandracluster_test.go (hunk 1 above)
Comment (Orange-cscetbon) [reply test/e2e/cassandracluster_test.go]: @jsanda did you miss my comment ☝️ ?
I propose to keep using `logrus` any not `log` here as we do in other files

Comment (jsanda) [reply test/e2e/cassandracluster_test.go]: I missed or forgot. You reminder is very timely though :) 

I am in the process of updating the test to be more consistent with other e2e tests. I am changing the test to create the cluster from the yaml file as is done in other tests. 

I will update the logging as well and hopefully commit/push shortly.

Comment (cscetbon) [reply test/e2e/cassandracluster_test.go]: Good to hear ! Thanks for you help again ;)

**Comment by jsanda** [comment]
@allamand I update the test in this PR to be more consistent with other tests and create the cluster object from yaml files. Can you please start/resume the CI pipeline?

**Comment by allamand** [comment]
LGTM, @Orange-cscetbon is it ok for you ?

**Pull request merged**